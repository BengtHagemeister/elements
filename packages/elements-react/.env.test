JEST_JUNIT_CLASSNAME="{filepath}"
JEST_JUNIT_OUTPUT_FILE="packages/elements-react/test-results/junit.xml"