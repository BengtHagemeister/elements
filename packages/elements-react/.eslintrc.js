// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

module.exports = {
  root: true,
  env: {
    node: true,
    browser: true,
    es2021: true,
  },
  extends: [
    "../../.eslintrc.cjs",
    "eslint:recommended",
    "plugin:react/recommended",
    "plugin:react/jsx-runtime",
  ],
  parserOptions: {
    ecmaVersion: 12,
    sourceType: "module",
    project: __dirname + "/tsconfig.json",
  },
  rules: {
    // false positive with typescript enums, and is covered by the typescript rule
    "no-unused-vars": "off",
    // covered by typescript
    "no-undef": "off",
    // TODO(jonas): define if we want to use this rule and if we want types or interfaces
    "@typescript-eslint/consistent-type-definitions": "off",
  },
  env: {
    jest: true,
  },
}
