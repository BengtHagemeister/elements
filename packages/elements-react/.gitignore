playwright/.cache/
playwright-report/*
.last-run.json