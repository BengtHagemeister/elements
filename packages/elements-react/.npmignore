playwright/.cache/
playwright-report/*
src/
playwright/
playwright-ct.config.ts
project.json
tsup.config.ts
tsconfig.test.json
!dist/