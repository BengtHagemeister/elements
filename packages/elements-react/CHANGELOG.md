## 1.0.0-next.10 (2024-10-14)


### 🩹 Fixes

- add missing translations in all places


### ❤️  Thank You

- Jonas Hungershausen

## 1.0.0-next.9 (2024-10-10)


### 🩹 Fixes

- add missing swedish translations


### ❤️  Thank You

- Jonas Hungershausen

## 1.0.0-next.8 (2024-10-07)


### 🩹 Fixes

- ⚠️  remove ory-default-theme class

- remove capitalize class from messages


#### ⚠️  Breaking Changes

- ⚠️  remove ory-default-theme class

### ❤️  Thank You

- Jonas Hungershausen

## 1.0.0-next.7 (2024-09-27)

### 🚀 Features

- add two-step login card

### 🩹 Fixes

- registration text in swedish translation

- show identifier_first node in card header

### ❤️ Thank You

- Jonas Hungershausen

## 1.0.0-next.6 (2024-09-18)

### 🩹 Fixes

- **intl:** export locales object from package root

### ❤️ Thank You

- Jonas Hungershausen

## 1.0.0-next.5 (2024-09-17)

### 🚀 Features

- @testing-library/react based tests

- **theme:** show resend code button on top of code input

### 🩹 Fixes

- **core:** use SDK instead of HTML form submissions

- **core:** use empty strings instead of undefined default values

- **core:** code form submission precendence

- ⚠️ **theme:** scope css classes with .ory-default-theme

- **theme:** social sign in button styling

- **intl:** use correct abbreveation for swedish locale

- **core:** passkey script discovery

#### ⚠️ Breaking Changes

- ⚠️ **theme:** scope css classes with .ory-default-theme

### ❤️ Thank You

- Jonas Hungershausen
- Miłosz
