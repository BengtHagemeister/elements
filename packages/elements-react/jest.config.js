// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

/** @type {import('ts-jest').JestConfigWithTsJest} **/
module.exports = {
  testEnvironment: "node",
  transform: {
    "^.+.tsx?$": ["ts-jest", {}],
  },
  testMatch: ["**/__tests__/**/*.(spec|test).[jt]s?(x)"],
}
