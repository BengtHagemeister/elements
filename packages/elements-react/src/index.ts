export type * from "./types"
export * from "./components"
export * from "./context"
export * from "./util"
