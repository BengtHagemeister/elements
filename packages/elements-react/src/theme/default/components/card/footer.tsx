// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { FlowType, UiNode, UiNodeInputAttributes } from "@ory/client-fetch"
import { useOryFlow } from "@ory/elements-react"
import { useFormContext } from "react-hook-form"
import { useIntl } from "react-intl"

export function DefaultCardFooter() {
  const { flowType } = useOryFlow()
  switch (flowType) {
    case FlowType.Login:
      return <LoginCardFooter />
    case FlowType.Registration:
      return <RegistrationCardFooter />
    case FlowType.Recovery:
      return <RecoveryCardFooter />
    case FlowType.Verification:
      return <VerificationCardFooter />
    default:
      return null
  }
}

function LoginCardFooter() {
  const { config, formState, initFlowUrl } = useOryFlow()
  const intl = useIntl()

  if (
    !config.project.registration_enabled ||
    formState.current !== "provide_identifier"
  ) {
    // The two-step login flow does not support the "navigation" between steps, so we don't have
    // anything to render on the footer in those steps
    return null
  }

  return (
    <span className="font-normal leading-normal antialiased text-interface-foreground-default-primary">
      {intl.formatMessage({
        id: "login.registration-label",
        defaultMessage: "No account?",
      })}{" "}
      <a
        className="text-button-link-brand-brand transition-colors hover:text-button-link-brand-brand-hover underline"
        href={initFlowUrl(FlowType.Registration)}
      >
        {intl.formatMessage({
          id: "login.registration-button",
          defaultMessage: "Sign up",
        })}
      </a>
    </span>
  )
}

function findScreenSelectionButton(
  nodes: UiNode[],
): { attributes: UiNodeInputAttributes } | undefined {
  return nodes.find(
    (node) =>
      node.attributes.node_type === "input" &&
      node.attributes.type === "submit" &&
      node.attributes.name === "screen",
  ) as { attributes: UiNodeInputAttributes }
}

function RegistrationCardFooter() {
  const intl = useIntl()
  const { flow, formState, initFlowUrl } = useOryFlow()
  if (formState.current === "select_method") {
    return null
  }
  const screenSelectionNode = findScreenSelectionButton(flow.ui.nodes)

  return (
    <span className="font-normal leading-normal antialiased">
      {formState.current === "method_active" ? (
        <>
          {screenSelectionNode && (
            <SelectScreenButton screenSelectionNode={screenSelectionNode} />
          )}
        </>
      ) : (
        <>
          {intl.formatMessage({
            id: "registration.login-label",
            defaultMessage: "Already have an account?",
          })}{" "}
          <a
            className="text-button-link-brand-brand transition-colors hover:text-button-link-brand-brand-hover underline"
            href={initFlowUrl(FlowType.Login)}
          >
            {intl.formatMessage({
              id: "registration.login-button",
              defaultMessage: "Sign in",
            })}
          </a>
        </>
      )}
    </span>
  )
}

type SelectScreenButtonProps = {
  screenSelectionNode: { attributes: UiNodeInputAttributes }
}

function SelectScreenButton({ screenSelectionNode }: SelectScreenButtonProps) {
  const { setValue } = useFormContext()
  const intl = useIntl()

  function handleScreenSelection() {
    setValue("method", "profile")
    if (screenSelectionNode) {
      setValue("screen", "credential-selection")
    }
  }
  return (
    <button
      className="font-medium text-button-link-brand-brand hover:text-button-link-brand-brand-hover"
      type="submit"
      onClick={handleScreenSelection}
    >
      {intl.formatMessage({
        id: "card.footer.select-another-method",
        defaultMessage: "Select another method",
      })}
    </button>
  )
}

function RecoveryCardFooter() {
  return null
}

function VerificationCardFooter() {
  return null
}
