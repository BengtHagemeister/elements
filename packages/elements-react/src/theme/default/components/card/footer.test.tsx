import { FlowType, LoginFlow } from "@ory/client-fetch"
import { render } from "@testing-library/react"
import { ComponentProps, ReactNode } from "react"
import { OryProvider } from "@ory/elements-react"
import { defaultConfiguration } from "../../../../tests/jest/test-utils"
import { DefaultCardFooter } from "./footer"

const flow = (returnTo?: string) => {
  return {
    request_url: `http://localhost:4455${returnTo ? "?return_to=" + returnTo : ""}`,
    state: "",
    ui: {
      action: "",
      method: "",
      nodes: [],
      messages: [],
    },
  } as unknown as LoginFlow
}

type renderProps = {
  providerProps: ComponentProps<typeof OryProvider>
}

const renderWithContext = (
  ui: ReactNode,
  { providerProps, ...renderOptions }: renderProps,
) => {
  return render(
    <OryProvider {...providerProps}>{ui}</OryProvider>,
    renderOptions,
  )
}

test("login footer should render return to in sign up link", () => {
  const x = renderWithContext(<DefaultCardFooter />, {
    providerProps: {
      components: null!,
      config: defaultConfiguration,
      flowType: FlowType.Login,
      flow: flow("https://ory.sh"),
    },
  })

  expect(x.container).toMatchSnapshot()
})

test("registration footer should render return to in sign in link", () => {
  const x = renderWithContext(<DefaultCardFooter />, {
    providerProps: {
      components: null!,
      config: defaultConfiguration,
      flowType: FlowType.Registration,
      flow: flow("https://ory.sh"),
    },
  })

  expect(x.container).toMatchSnapshot()
})

test("login footer should render sign up link", () => {
  const x = renderWithContext(<DefaultCardFooter />, {
    providerProps: {
      components: null!,
      config: defaultConfiguration,
      flowType: FlowType.Login,
      flow: flow(),
    },
  })

  expect(x.container).toMatchSnapshot()
})
test("registration footer should render sign in link", () => {
  const x = renderWithContext(<DefaultCardFooter />, {
    providerProps: {
      components: null!,
      config: defaultConfiguration,
      flowType: FlowType.Registration,
      flow: flow(),
    },
  })

  expect(x.container).toMatchSnapshot()
})
