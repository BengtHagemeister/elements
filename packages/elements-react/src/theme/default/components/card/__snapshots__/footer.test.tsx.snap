// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login footer should render return to in sign up link 1`] = `
<div>
  <span
    class="font-normal leading-normal antialiased text-interface-foreground-default-primary"
  >
    Don't have an account?
     
    <a
      class="text-button-link-brand-brand transition-colors hover:text-button-link-brand-brand-hover underline"
      href="http://localhost:4455/self-service/registration/browser?return_to=https://ory.sh"
    >
      Sign up
    </a>
  </span>
</div>
`;

exports[`login footer should render sign up link 1`] = `
<div>
  <span
    class="font-normal leading-normal antialiased text-interface-foreground-default-primary"
  >
    Don't have an account?
     
    <a
      class="text-button-link-brand-brand transition-colors hover:text-button-link-brand-brand-hover underline"
      href="http://localhost:4455/self-service/registration/browser"
    >
      Sign up
    </a>
  </span>
</div>
`;

exports[`registration footer should render return to in sign in link 1`] = `
<div>
  <span
    class="font-normal leading-normal antialiased"
  >
    Already have an account?
     
    <a
      class="text-button-link-brand-brand transition-colors hover:text-button-link-brand-brand-hover underline"
      href="http://localhost:4455/self-service/login/browser?return_to=https://ory.sh"
    >
      Sign in
    </a>
  </span>
</div>
`;

exports[`registration footer should render sign in link 1`] = `
<div>
  <span
    class="font-normal leading-normal antialiased"
  >
    Already have an account?
     
    <a
      class="text-button-link-brand-brand transition-colors hover:text-button-link-brand-brand-hover underline"
      href="http://localhost:4455/self-service/login/browser"
    >
      Sign in
    </a>
  </span>
</div>
`;
